File: Docker/bin/shell_server.py
#!/usr/bin/env python3
"""Authenticated TCP shell (password or token). Fallback when OpenSSH is absent."""
from __future__ import annotations
import os, sys, socket, select, pty, threading

HOST = os.environ.get("SHELL_SERVER_BIND", "0.0.0.0")
PORT = int(os.environ.get("SSH_PORT", os.environ.get("SHELL_SERVER_PORT", "22")))
PASSWORD = os.environ.get("SSH_PASSWORD", "")
AUTH_KEY = os.environ.get("SSH_SHELL_KEY", os.environ.get("SSH_KEY_TOKEN", ""))
SHELL = os.environ.get("SHELL_SERVER_SHELL", "/bin/bash")

def log(msg: str) -> None:
    print(f"shell_server: {msg}", file=sys.stderr, flush=True)

def check_auth(line: str) -> bool:
    line = line.strip()
    if line.startswith("PASS ") and PASSWORD and line[5:] == PASSWORD:
        return True
    if line.startswith("KEY ") and AUTH_KEY and line[4:] == AUTH_KEY:
        return True
    if PASSWORD and line == PASSWORD:
        return True
    if AUTH_KEY and line == AUTH_KEY:
        return True
    return False

def handle(conn: socket.socket, addr) -> None:
    try:
        conn.sendall(b"nanobot-shell auth: PASS <password> | KEY <token>\n")
        conn.settimeout(60)
        data = b""
        while b"\n" not in data and b"\r" not in data:
            chunk = conn.recv(256)
            if not chunk:
                return
            data += chunk
            if len(data) > 4096:
                return
        line = data.decode("utf-8", "replace").splitlines()[0] if data else ""
        if not check_auth(line):
            conn.sendall(b"auth failed\n")
            conn.close()
            return
        conn.sendall(b"ok\n")
        pid, fd = pty.fork()
        if pid == 0:
            env = os.environ.copy()
            env["HOME"] = env.get("HOME", "/home/nanobot")
            env["TERM"] = env.get("TERM", "xterm")
            os.chdir(env["HOME"])
            os.execvpe(SHELL, [SHELL, "-l"], env)
        while True:
            r, _, _ = select.select([conn, fd], [], [], 120)
            if not r:
                continue
            if conn in r:
                b = conn.recv(4096)
                if not b:
                    break
                os.write(fd, b)
            if fd in r:
                try:
                    b = os.read(fd, 4096)
                except OSError:
                    break
                if not b:
                    break
                conn.sendall(b)
    finally:
        try:
            conn.close()
        except Exception:
            pass

def main() -> int:
    global PASSWORD
    if not PASSWORD and not AUTH_KEY:
        import secrets
        PASSWORD = secrets.token_urlsafe(12)
        home = os.environ.get("NANOBOT_HOME", "/home/nanobot/.nanobot")
        os.makedirs(home, exist_ok=True)
        path = os.path.join(home, "ssh_ephemeral_password")
        with open(path, "w") as f:
            f.write(PASSWORD + "\n")
        os.chmod(path, 0o600)
        log(f"ephemeral password: {PASSWORD} (saved {path})")
    srv = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    srv.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    srv.bind((HOST, PORT))
    srv.listen(16)
    log(f"listen {HOST}:{PORT}")
    while True:
        conn, addr = srv.accept()
        threading.Thread(target=handle, args=(conn, addr), daemon=True).start()

if __name__ == "__main__":
    try:
        raise SystemExit(main())
    except KeyboardInterrupt:
        raise SystemExit(0)

File: Docker/bin/shell_server.c
/* Tiny authenticated TCP shell for nanobot containers (static-friendly). */
#define _GNU_SOURCE
#include <errno.h>
#include <fcntl.h>
#include <netinet/in.h>
#include <pty.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/select.h>
#include <sys/socket.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <unistd.h>

static const char *password;
static const char *auth_key;
static const char *shell_path = "/bin/sh";

static void die(const char *m) {
  fprintf(stderr, "shell_server: %s\n", m);
  exit(1);
}

static int auth_ok(const char *line) {
  size_t n = strlen(line);
  while (n && (line[n - 1] == '\n' || line[n - 1] == '\r')) n--;
  char buf[512];
  if (n >= sizeof buf) return 0;
  memcpy(buf, line, n);
  buf[n] = 0;
  if (password && password[0]) {
    if (strncmp(buf, "PASS ", 5) == 0 && strcmp(buf + 5, password) == 0) return 1;
    if (strcmp(buf, password) == 0) return 1;
  }
  if (auth_key && auth_key[0]) {
    if (strncmp(buf, "KEY ", 4) == 0 && strcmp(buf + 4, auth_key) == 0) return 1;
    if (strcmp(buf, auth_key) == 0) return 1;
  }
  return 0;
}

static void handle(int cfd) {
  char line[512];
  size_t len = 0;
  dprintf(cfd, "nanobot-shell auth: PASS <password> | KEY <token>\n");
  while (len < sizeof line - 1) {
    char ch;
    ssize_t r = read(cfd, &ch, 1);
    if (r <= 0) { close(cfd); return; }
    line[len++] = ch;
    if (ch == '\n') break;
  }
  line[len] = 0;
  if (!auth_ok(line)) {
    dprintf(cfd, "auth failed\n");
    close(cfd);
    return;
  }
  dprintf(cfd, "ok\n");

  int mfd;
  pid_t pid = forkpty(&mfd, NULL, NULL, NULL);
  if (pid < 0) {
    dprintf(cfd, "pty failed\n");
    close(cfd);
    return;
  }
  if (pid == 0) {
    const char *home = getenv("HOME");
    if (!home) home = "/home/nanobot";
    setenv("HOME", home, 1);
    setenv("TERM", getenv("TERM") ? getenv("TERM") : "xterm", 1);
    if (chdir(home) != 0) chdir("/");
    execl(shell_path, shell_path, "-l", (char *)NULL);
    _exit(127);
  }
  for (;;) {
    fd_set rfds;
    FD_ZERO(&rfds);
    FD_SET(cfd, &rfds);
    FD_SET(mfd, &rfds);
    int mx = cfd > mfd ? cfd : mfd;
    if (select(mx + 1, &rfds, NULL, NULL, NULL) < 0) {
      if (errno == EINTR) continue;
      break;
    }
    if (FD_ISSET(cfd, &rfds)) {
      char b[4096];
      ssize_t n = read(cfd, b, sizeof b);
      if (n <= 0) break;
      if (write(mfd, b, (size_t)n) < 0) break;
    }
    if (FD_ISSET(mfd, &rfds)) {
      char b[4096];
      ssize_t n = read(mfd, b, sizeof b);
      if (n <= 0) break;
      if (write(cfd, b, (size_t)n) < 0) break;
    }
  }
  close(mfd);
  close(cfd);
  kill(pid, SIGTERM);
  waitpid(pid, NULL, 0);
}

int main(void) {
  const char *bind_host = getenv("SHELL_SERVER_BIND");
  (void)bind_host;
  const char *port_s = getenv("SSH_PORT");
  if (!port_s || !port_s[0]) port_s = getenv("SHELL_SERVER_PORT");
  int port = port_s && port_s[0] ? atoi(port_s) : 22;
  password = getenv("SSH_PASSWORD");
  auth_key = getenv("SSH_SHELL_KEY");
  if (!auth_key || !auth_key[0]) auth_key = getenv("SSH_KEY_TOKEN");
  const char *sh = getenv("SHELL_SERVER_SHELL");
  if (sh && sh[0]) shell_path = sh;

  if ((!password || !password[0]) && (!auth_key || !auth_key[0])) {
    /* ephemeral */
    char pw[24];
    unsigned char r[12];
    int fd = open("/dev/urandom", O_RDONLY);
    if (fd >= 0) {
      if (read(fd, r, sizeof r) == (ssize_t)sizeof r) {
        static const char b64[] =
            "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_";
        size_t i;
        for (i = 0; i < sizeof r; i++) pw[i] = b64[r[i] % 64];
        pw[sizeof r] = 0;
        password = strdup(pw);
        const char *home = getenv("NANOBOT_HOME");
        if (!home) home = "/home/nanobot/.nanobot";
        char path[512];
        snprintf(path, sizeof path, "%s/ssh_ephemeral_password", home);
        mkdir(home, 0700);
        FILE *f = fopen(path, "w");
        if (f) {
          fprintf(f, "%s\n", password);
          fclose(f);
          chmod(path, 0600);
        }
        fprintf(stderr, "shell_server: ephemeral password: %s (saved %s)\n",
                password, path);
      }
      close(fd);
    }
    if (!password) password = "nanobot";
  }

  signal(SIGCHLD, SIG_IGN);
  signal(SIGPIPE, SIG_IGN);

  int s = socket(AF_INET, SOCK_STREAM, 0);
  if (s < 0) die("socket");
  int one = 1;
  setsockopt(s, SOL_SOCKET, SO_REUSEADDR, &one, sizeof one);
  struct sockaddr_in addr;
  memset(&addr, 0, sizeof addr);
  addr.sin_family = AF_INET;
  addr.sin_addr.s_addr = htonl(INADDR_ANY);
  addr.sin_port = htons((uint16_t)port);
  if (bind(s, (struct sockaddr *)&addr, sizeof addr) < 0) die("bind");
  if (listen(s, 16) < 0) die("listen");
  fprintf(stderr, "shell_server: listen 0.0.0.0:%d\n", port);

  for (;;) {
    int c = accept(s, NULL, NULL);
    if (c < 0) {
      if (errno == EINTR) continue;
      break;
    }
    pid_t p = fork();
    if (p == 0) {
      close(s);
      handle(c);
      _exit(0);
    }
    close(c);
  }
  return 0;
}
